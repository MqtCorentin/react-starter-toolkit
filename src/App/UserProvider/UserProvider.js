import React, { createContext } from 'react';
import { get, isEmpty } from 'lodash';
import { useReactOidc } from '@axa-fr/react-oidc-context';
import { PROFILS } from 'shared/constants';

export const UserContext = createContext({
  authName: '',
  authAccessToken: '',
  authRole: '',
  authUid: '',
});
export const UserContextProvider = UserContext.Provider;

const NON_CONNECTE = 'Non ConnectÃ©';

export const getAuthName = ({ oidcUser, isEmptyFn = isEmpty, getFn = get }) =>
  !isEmptyFn(getFn(oidcUser, 'profile.name')) ? oidcUser.profile.name : NON_CONNECTE;

export const getAuthAccessToken = ({ oidcUser, isEmptyFn = isEmpty, getFn = get }) =>
  !isEmptyFn(getFn(oidcUser, 'access_token')) ? oidcUser.access_token : '';

export const setAuthRole = ({ memberOf, profils = PROFILS }) => profils.map(profil => (memberOf.search(`${profil}`) !== -1 ? profil : '')).join('');

export const getAuthRole = ({ oidcUser, setAuthRoleFn = setAuthRole, isEmptyFn = isEmpty, getFn = get }) =>
  !isEmptyFn(getFn(oidcUser, 'profile.member_of')) ? setAuthRoleFn({ memberOf: oidcUser.profile.member_of[0] }) : '';

export const getAuthUid = ({ oidcUser, isEmptyFn = isEmpty, getFn = get }) =>
  !isEmptyFn(getFn(oidcUser, 'profile.axa_uid_racf')) ? oidcUser.profile.axa_uid_racf : '';

/**
 * MAAM gives us : "member_of": [ELMU_Admin,CN=IAM_ELMU,OU=applis,O=axafrance,DC=REWACAD,DC=axa-fr,DC=intraxa]
 * @param {Object} oidcUser
 */
const extractDataFromOAuthToken = (
  oidcUser,
  getAuthNameFn = getAuthName,
  getAuthAccessTokenFn = getAuthAccessToken,
  getAuthRoleFn = getAuthRole,
  getAuthUidFn = getAuthUid,
) => ({
  authName: getAuthNameFn({ oidcUser }),
  authAccessToken: getAuthAccessTokenFn({ oidcUser }),
  authRole: getAuthRoleFn({ oidcUser }),
  authUid: getAuthUidFn({ oidcUser }),
});

const UserProvider = ({ children, useReactOidcFn = useReactOidc }) => {
  const { oidcUser } = useReactOidcFn();
  return <UserContextProvider value={{ ...extractDataFromOAuthToken(oidcUser) }}>{children}</UserContextProvider>;
};
export default UserProvider;
