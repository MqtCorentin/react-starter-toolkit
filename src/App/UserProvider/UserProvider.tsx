import { createContext, ReactNode } from 'react';
import { useOidcUser } from '@axa-fr/react-oidc-context';
import { isEmpty } from 'lodash';
import { PROFILS } from 'shared/constants';

export const UserContext = createContext({
  authName: '',
  authRole: '',
  authUid: '',
  isEnabled: true,
  isLoading: true,
});
UserContext.displayName = 'UserContext';

const NON_CONNECTE = 'Non ConnectÃ©';

type ToidcUser = {
  member_of?: string[];
  name: string;
  axa_uid_racf?: string;
};

/**
 * getAuthName
 */

type TgetAuthName = {
  oidcUser: ToidcUser;
  isEmptyFn?: typeof isEmpty;
};

export const getAuthName = ({ oidcUser, isEmptyFn = isEmpty }: TgetAuthName) => (!isEmptyFn(oidcUser?.name ?? '') ? oidcUser.name : NON_CONNECTE);

/**
 * setAuthRole
 */
type TsetAuthRole = {
  memberOf: string;
  profils?: typeof PROFILS;
};

export const setAuthRole = ({ memberOf, profils = PROFILS }: TsetAuthRole) =>
  profils.map(profil => (memberOf.search(`${profil}`) !== -1 ? profil : '')).join('');

/**
 * getAuthRole
 */

type TgetAuthRole = {
  oidcUser: ToidcUser;
  setAuthRoleFn?: typeof setAuthRole;
};

export const getAuthRole = ({ oidcUser, setAuthRoleFn = setAuthRole }: TgetAuthRole) => setAuthRoleFn({ memberOf: oidcUser?.member_of?.[0] ?? '' });

/**
 * getAuthUid
 */

type TgetAuthUid = {
  oidcUser: ToidcUser;
  isEmptyFn?: typeof isEmpty;
};

export const getAuthUid = ({ oidcUser, isEmptyFn = isEmpty }: TgetAuthUid) =>
  (!isEmptyFn(oidcUser?.axa_uid_racf ?? '') ? oidcUser.axa_uid_racf : '') || '';

/**
 * extractDataFromOAuthToken
 * @param {Object} oidcUser
 */

type TextractDataFromOAuthToken = {
  oidcUser: ToidcUser;
  getAuthNameFn?: typeof getAuthName;
  getAuthRoleFn?: typeof getAuthRole;
  getAuthUidFn?: typeof getAuthUid;
};

const extractDataFromOAuthToken = ({
  oidcUser,
  getAuthNameFn = getAuthName,
  getAuthRoleFn = getAuthRole,
  getAuthUidFn = getAuthUid,
}: TextractDataFromOAuthToken) => ({
  authName: getAuthNameFn({ oidcUser }),
  authRole: getAuthRoleFn({ oidcUser }),
  authUid: getAuthUidFn({ oidcUser }),
  isLoading: !oidcUser,
});

/**
 * UserProvider
 * @param {Function} useOidcUser
 * @param {JSX} children
 */

type TUserProvider = {
  children: ReactNode;
  useOidcUserFn: typeof useOidcUser;
  extractDataFromOAuthTokenFn?: typeof extractDataFromOAuthToken;
  isEnabled?: boolean;
};

const UserProvider = ({
  children,
  useOidcUserFn,
  extractDataFromOAuthTokenFn = extractDataFromOAuthToken,
  isEnabled = true,
  ...rest
}: TUserProvider) => {
  const { oidcUser } = useOidcUserFn();
  return <UserContext.Provider value={{ ...extractDataFromOAuthTokenFn({ oidcUser }), isEnabled, ...rest }}>{children}</UserContext.Provider>;
};

export default UserProvider;
