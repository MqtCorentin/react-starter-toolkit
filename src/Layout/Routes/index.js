export { default, WithAuth } from './Routes';
