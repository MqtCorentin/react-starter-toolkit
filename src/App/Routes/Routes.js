import React, { useContext } from 'react';
import { func } from 'prop-types';
import { Routes, Route, Navigate } from 'react-router-dom';
import PageNotFound from 'pages/NotFound';
import PageUnauthorize from 'pages/Unauthorize';
import Members from 'pages/Members';
import SearchMembers from 'pages/SearchMembers';
import SlashDesignSystem from 'pages/SlashDesignSystem';
import Modal from 'pages/Demos/Modal';
import Button from 'pages/Demos/Button';
import Dashboard from 'pages/Dashboard';
import ROUTE_URL from 'Layout/constants';
import { UserContext } from 'App/UserProvider';

export const WithAuth = ({ Component, NavigateCmpt = Navigate, authorized = [''], UserContextObj = UserContext }) => {
  const userContext = useContext(UserContextObj);
  return authorized.includes(userContext?.authRole) || !userContext.isEnabled ? <Component /> : <NavigateCmpt to={ROUTE_URL.UNAUTHORIZE} />;
};

const RoutesCmpt = ({
  SlashDesignSystemCmpt,
  DashboardCmpt,
  MembersCmpt,
  SearchMembersCmpt,
  ModalCmpt,
  ButtonCmpt,
  PageUnauthorizeCmpt,
  WithAuthCmpt,
}) => (
  <Routes>
    <Route exact path={ROUTE_URL.SLASH} element={<WithAuthCmpt Component={SlashDesignSystemCmpt} />} />
    <Route exact path={ROUTE_URL.MEMBERS} element={<WithAuthCmpt Component={MembersCmpt} />} />
    <Route exact path={ROUTE_URL.SEARCHMEMBERS} element={<WithAuthCmpt Component={SearchMembersCmpt} />} />
    <Route exact path={ROUTE_URL.MODAL} element={<WithAuthCmpt Component={ModalCmpt} />} />
    <Route exact path={ROUTE_URL.BUTTON} element={<WithAuthCmpt Component={ButtonCmpt} />} />
    <Route exact path={ROUTE_URL.DASHBOARD} element={<WithAuthCmpt Component={DashboardCmpt} />} />
    <Route exact path={ROUTE_URL.UNAUTHORIZE} element={<PageUnauthorizeCmpt />} />
    <Route path="*" element={<PageNotFound />} />
  </Routes>
);

RoutesCmpt.propTypes = {
  SlashDesignSystemCmpt: func,
  DashboardCmpt: func,
  MembersCmpt: func,
  SearchMembersCmpt: func,
  ModalCmpt: func,
  ButtonCmpt: func,
  PageUnauthorizeCmpt: func,
  WithAuthCmpt: func,
};

RoutesCmpt.defaultProps = {
  SlashDesignSystemCmpt: SlashDesignSystem,
  DashboardCmpt: Dashboard,
  MembersCmpt: Members,
  SearchMembersCmpt: SearchMembers,
  ModalCmpt: Modal,
  ButtonCmpt: Button,
  PageUnauthorizeCmpt: PageUnauthorize,
  WithAuthCmpt: WithAuth,
};

export default RoutesCmpt;
