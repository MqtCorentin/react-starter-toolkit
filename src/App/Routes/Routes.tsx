import { useContext, ComponentProps } from 'react';
import { Routes, Route, Navigate } from 'react-router-dom';
import PageNotFound from 'pages/NotFound';
import PageUnauthorize from 'pages/Unauthorize';
import Members from 'pages/Members';
import SearchMembers from 'pages/SearchMembers';
import SlashDesignSystem from 'pages/SlashDesignSystem';
import Modal from 'pages/Demos/Modal';
import Button from 'pages/Demos/Button';
import Notification from 'pages/Demos/Notification';
import Dashboard from 'pages/Dashboard';
import ROUTE_URL from 'App/Routes/constants';
import { UserContext } from 'App/UserProvider';
import Loader, { MODES } from 'shared/components/Loader';

export const withAuth = <T extends object>(
  Component: React.ComponentType<T>,
  UserContextObj = UserContext,
  authorized = [''],
  NavigateCmpt = Navigate,
  LoaderCmpt = Loader,
) => {
  const NewComp = (props: ComponentProps<typeof Component> | object) => {
    const userContext = useContext(UserContextObj);

    if (userContext.isEnabled && userContext.isLoading) {
      return <LoaderCmpt text="Chargement des donnÃ©es utilisateur..." mode={MODES.get} classModifier="fullscreen" />;
    }

    return authorized.includes(userContext?.authRole) || !userContext.isEnabled ? (
      <Component {...(props as T)} />
    ) : (
      <NavigateCmpt to={ROUTE_URL.UNAUTHORIZE} />
    );
  };

  return <NewComp />;
};

type TRoutesCmpt = {
  SlashDesignSystemCmpt?: typeof SlashDesignSystem;
  DashboardCmpt?: typeof Dashboard;
  MembersCmpt?: typeof Members;
  SearchMembersCmpt?: typeof SearchMembers;
  ModalCmpt?: typeof Modal;
  ButtonCmpt?: typeof Button;
  NotificationCmpt?: typeof Notification;
  PageUnauthorizeCmpt?: typeof PageUnauthorize;
  withAuthFn?: typeof withAuth;
};

const RoutesCmpt = ({
  SlashDesignSystemCmpt = SlashDesignSystem,
  DashboardCmpt = Dashboard,
  MembersCmpt = Members,
  SearchMembersCmpt = SearchMembers,
  ModalCmpt = Modal,
  ButtonCmpt = Button,
  NotificationCmpt = Notification,
  PageUnauthorizeCmpt = PageUnauthorize,
  withAuthFn = withAuth,
}: TRoutesCmpt) => (
  <Routes>
    <Route path={ROUTE_URL.SLASH} element={withAuthFn(SlashDesignSystemCmpt)} />
    <Route path={ROUTE_URL.MEMBERS} element={withAuthFn(MembersCmpt)} />
    <Route path={ROUTE_URL.SEARCHMEMBERS} element={withAuthFn(SearchMembersCmpt)} />
    <Route path={ROUTE_URL.MODAL} element={withAuthFn(ModalCmpt)} />
    <Route path={ROUTE_URL.BUTTON} element={withAuthFn(ButtonCmpt)} />
    <Route path={ROUTE_URL.NOTIFICATION} element={withAuthFn(NotificationCmpt)} />
    <Route path={ROUTE_URL.DASHBOARD} element={withAuthFn(DashboardCmpt)} />
    <Route path={ROUTE_URL.UNAUTHORIZE} element={<PageUnauthorizeCmpt />} />
    <Route path="*" element={<PageNotFound />} />
  </Routes>
);

export default RoutesCmpt;
