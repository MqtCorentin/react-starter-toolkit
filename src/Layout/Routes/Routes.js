import React, { useContext } from 'react';
import { func } from 'prop-types';
import { Routes, Route, Navigate } from 'react-router-dom';
import PageNotFound from 'NotFound';
import PageUnauthorize from 'Unauthorize';
import Members from 'Members';
import SearchMembers from 'SearchMembers';
import SlashDesignSystem from 'SlashDesignSystem';
import Modal from 'Demos/Modal';
import Button from 'Demos/Button';
import Dashboard from 'Dashboard';
import ROUTE_URL from 'Layout/constants';
import { UserContext } from 'App/UserProvider';

export const WithAuth = ({ Component, NavigateCmpt = Navigate, authorized = [''], UserContextObj = UserContext }) => {
  const userContext = useContext(UserContextObj);
  return authorized.includes(userContext?.authRole) || !userContext.isEnabled ? <Component /> : <NavigateCmpt to={ROUTE_URL.UNAUTHORIZE} />;
};

const RoutesCmpt = ({
  SlashDesignSystemCmpt,
  DashboardCmpt,
  MembersCmpt,
  SearchMembersCmpt,
  ModalCmpt,
  ButtonCmpt,
  PageUnauthorizeCmpt,
  WithAuthCmpt,
}) => (
  <Routes>
    <Route exact path={ROUTE_URL.SLASH} element={<WithAuthCmpt Component={SlashDesignSystemCmpt} />} />
    <Route exact path={ROUTE_URL.MEMBERS} element={<WithAuthCmpt Component={MembersCmpt} />} />
    <Route exact path={ROUTE_URL.SEARCHMEMBERS} element={<WithAuthCmpt Component={SearchMembersCmpt} />} />
    <Route exact path={ROUTE_URL.MODAL} element={<WithAuthCmpt Component={ModalCmpt} />} />
    <Route exact path={ROUTE_URL.BUTTON} element={<WithAuthCmpt Component={ButtonCmpt} />} />
    <Route exact path={ROUTE_URL.DASHBOARD} element={<WithAuthCmpt Component={DashboardCmpt} />} />
    <Route exact path={ROUTE_URL.UNAUTHORIZE} element={<PageUnauthorizeCmpt />} />
    <Route path="*" element={<PageNotFound />} />
  </Routes>
);

RoutesCmpt.propTypes = {
  SlashDesignSystemCmpt: func,
  DashboardCmpt: func,
  MembersCmpt: func,
  SearchMembersCmpt: func,
  ModalCmpt: func,
  ButtonCmpt: func,
  PageUnauthorizeCmpt: func,
  WithAuthCmpt: func,
};

RoutesCmpt.defaultProps = {
  SlashDesignSystemCmpt: SlashDesignSystem,
  DashboardCmpt: Dashboard,
  MembersCmpt: Members,
  SearchMembersCmpt: SearchMembers,
  ModalCmpt: Modal,
  ButtonCmpt: Button,
  PageUnauthorizeCmpt: PageUnauthorize,
  WithAuthCmpt: WithAuth,
};

export default RoutesCmpt;
