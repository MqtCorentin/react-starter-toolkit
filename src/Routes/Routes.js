import React from 'react';
import { Switch, Route } from 'react-router-dom';

import PageNotFound from 'NotFound';
import Members from 'Members';
import Layout from 'Layout';

export const renderLayoutMembers = props => Layout(Members, props);

const Routes = () => (
  <Switch>
    <Route exact path="/" render={renderLayoutMembers} />
    <Route component={PageNotFound} />
  </Switch>
);

export default Routes;
