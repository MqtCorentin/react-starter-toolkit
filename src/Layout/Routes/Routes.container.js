import React, { useContext } from 'react';
import { UserContext } from 'App/User';
import Routes from './Routes';

const RouteWithUserInfo = () => {
  const userContext = useContext(UserContext);
  return <Routes {...userContext} />;
};

export default RouteWithUserInfo;
