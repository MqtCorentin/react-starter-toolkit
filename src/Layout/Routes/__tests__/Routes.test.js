import React, { createContext } from 'react';
import { MemoryRouter } from 'react-router-dom';
import { render } from '@testing-library/react';
import UserProvider from 'App/UserProvider';
import { renderWithWrapperStaticRouter } from 'shared/testsUtils';
import Routes, { WithAuth } from '..';

const defaultPropsMock = {
  SlashDesignSystemCmpt: () => <div>SlashDesignSystemCmpt</div>,
  MembersCmpt: () => <div>MembersCmpt</div>,
  SearchMembersCmpt: () => <div>SearchMembersCmpt</div>,
  DashboardCmpt: () => <div>DashboardCmpt</div>,
  ModalCmpt: () => <div>ModalCmpt</div>,
  ButtonCmpt: () => <div>ButtonCmpt</div>,
  PageUnauthorizeCmpt: () => <div>PageUnauthorizeCmpt</div>,
};

const renderRoute = ({ role, name, route, defaultProps = defaultPropsMock }) => {
  const useOidcUser = jest.fn().mockReturnValue({
    oidcUser: {
      name,
      profile: {
        member_of: [`CN=${role}`],
      },
    },
  });

  return render(
    <MemoryRouter initialEntries={[route]}>
      <UserProvider useOidcUser={useOidcUser} isEnabled>
        <Routes {...defaultProps} />
      </UserProvider>
    </MemoryRouter>,
  );
};

describe('<Routes />', () => {
  it.each`
    role  | name        | route
    ${''} | ${'Samuel'} | ${'/'}
    ${''} | ${'Samuel'} | ${'/members'}
    ${''} | ${'Samuel'} | ${'/dashboard'}
    ${''} | ${'Samuel'} | ${'/search-members'}
    ${''} | ${'Samuel'} | ${'/demos/modal'}
    ${''} | ${'Samuel'} | ${'/demos/button'}
    ${''} | ${'Samuel'} | ${'/forbidden'}
  `('Should render page when user profil is authorized, role: $role, name: $name, route: $route', ({ role, name, route }) => {
    const { asFragment } = renderRoute({ role, name, route });
    expect(asFragment()).toMatchSnapshot();
  });

  it.each`
    role                   | name     | route
    ${'user unauthorized'} | ${'sam'} | ${'/'}
    ${'user unauthorized'} | ${'sam'} | ${'/members'}
    ${'user unauthorized'} | ${'sam'} | ${'/dashboard'}
    ${'user unauthorized'} | ${'sam'} | ${'/search-members'}
    ${'user unauthorized'} | ${'sam'} | ${'/demos/modal'}
    ${'user unauthorized'} | ${'sam'} | ${'/demos/button'}
  `('Should render forbidden page when user profil is unauthorized, authRole: $authRole, route: $route', ({ role, name, route }) => {
    const { getByText } = renderRoute({ role, name, route, defaultProps: { WithAuthCmpt: props => <WithAuth {...props} authorized={['admin']} /> } });
    expect(getByText('403')).toBeInTheDocument();
  });

  it.each`
    route
    ${'/no-exist-route'}
  `('Should render 404 when election route: $route is not correct', ({ route }) => {
    const { getByText } = renderRoute({ route, role: '', name: 'samuel', defaultProps: {} });
    expect(getByText('404')).toBeInTheDocument();
  });
});

describe('Render WithAuth', () => {
  it.each`
    authRole | authName
    ${''}    | ${'Samuel'}
  `('Should render RouteCmpt when user profile is authorized (authRole, authName)', ({ authRole, authName }) => {
    const UserContextObjMock = createContext({ authRole, authName, isEnabled: true });
    const Component = () => <p>component</p>;
    const { getByText } = renderWithWrapperStaticRouter(<WithAuth UserContextObj={UserContextObjMock} Component={Component} />);
    expect(getByText('component')).toBeInTheDocument();
  });

  it('Should render NavigateCmpt when user profile is unauthorized', () => {
    const UserContextObjMock = createContext({ isEnabled: true });
    const NavigateCmpt = () => <p>redirect component</p>;
    const { getByText } = renderWithWrapperStaticRouter(<WithAuth UserContextObj={UserContextObjMock} NavigateCmpt={NavigateCmpt} />);
    expect(getByText('redirect component')).toBeInTheDocument();
  });
});
