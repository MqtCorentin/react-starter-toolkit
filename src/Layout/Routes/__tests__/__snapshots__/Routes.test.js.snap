// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<Routes /> Should render page when user profil is authorized, role: , name: Samuel, route: / 1`] = `
<DocumentFragment>
  <div>
    SlashDesignSystemCmpt
  </div>
</DocumentFragment>
`;

exports[`<Routes /> Should render page when user profil is authorized, role: , name: Samuel, route: /dashboard 1`] = `
<DocumentFragment>
  <div>
    DashboardCmpt
  </div>
</DocumentFragment>
`;

exports[`<Routes /> Should render page when user profil is authorized, role: , name: Samuel, route: /demos/button 1`] = `
<DocumentFragment>
  <div>
    ButtonCmpt
  </div>
</DocumentFragment>
`;

exports[`<Routes /> Should render page when user profil is authorized, role: , name: Samuel, route: /demos/modal 1`] = `
<DocumentFragment>
  <div>
    ModalCmpt
  </div>
</DocumentFragment>
`;

exports[`<Routes /> Should render page when user profil is authorized, role: , name: Samuel, route: /forbidden 1`] = `
<DocumentFragment>
  <div>
    PageUnauthorizeCmpt
  </div>
</DocumentFragment>
`;

exports[`<Routes /> Should render page when user profil is authorized, role: , name: Samuel, route: /members 1`] = `
<DocumentFragment>
  <div>
    MembersCmpt
  </div>
</DocumentFragment>
`;

exports[`<Routes /> Should render page when user profil is authorized, role: , name: Samuel, route: /search-members 1`] = `
<DocumentFragment>
  <div>
    SearchMembersCmpt
  </div>
</DocumentFragment>
`;
